sbi/diagnostics/tarp.py

```diff
@@ -0,0 +1,314 @@
+        # We use the solution proposed here: https://stackoverflow.com/a/61689175
+        # to update the pbar only after the workers finished a task.
+        with tqdm_joblib(
+            tqdm(
+                xs_batches,
+                disable=not show_progress_bar,
+                desc=f"Performing {num_sim_samples} posterior runs in"
+                f"{len(xs_batches)} batches.",
+                total=len(xs_batches),
+            )
+        ) as _:
+            samples: Tensor
+            samples = Parallel(n_jobs=num_workers)(  # pyright: ignore[reportAssignmentType]
+                delayed(infer_posterior_on_batch)(
+                    xs_batch, posterior, num_posterior_samples
+                )
+                for xs_batch in xs_batches
+            )
+    else:
+        pbar = tqdm(
+            total=num_sim_samples,
+            disable=not show_progress_bar,
+            desc=f"Running {num_sim_samples} samples for tarp analysis.",
+        )
+
+        with pbar:
+            samples = []
+            for xs_batch in xs_batches:
+                samples.append(
+                    infer_posterior_on_batch(xs_batch, posterior, num_posterior_samples)
+                )
+                pbar.update(infer_batch_size)
+            samples = torch.cat(samples, dim=1)
+
+    return samples
+
+
+def _check_references(references: Tensor, theta: Tensor) -> Tensor:
+    """construct references"""
+
+    num_dims = theta.shape[-1]
+    num_sims = theta.shape[0]
+
+    if not isinstance(references, Tensor):
+        # obtain min/max per dimension of theta
+        lo = (
+            torch.min(theta, dim=-2).values.min(axis=0).values
+        )  # should be 0 if normalized
+        hi = (
+            torch.max(theta, dim=-2).values.max(axis=0).values
+        )  # should be 1 if normalized
+
+        refpdf = torch.distributions.Uniform(low=lo, high=hi)
+        # sample one reference point for each entry in theta
+        references = refpdf.sample((1, num_sims))
+    else:
+        if len(references.shape) == 2:
+            # add singleton dimension in front
+            references = references.unsqueeze(0)
+
+        if len(references.shape) == 3 and references.shape[0] != 1:
+            raise ValueError(
+                f"""references must be a 2D array with a singular first
+                    dimension, received {references.shape}"""
+            )
+
+        if references.shape[-2] != num_sims:
+            raise ValueError(
+                f"references must have the same number samples as samples,"
+                f"received {references.shape[-2]} != {num_sims}"
+            )
+
+        if references.shape[-1] != num_dims:
+            raise ValueError(
+                "references must have the same number of dimensions as "
+                f"samples or theta, received {references.shape[-1]}"
+                f"!= {num_dims}"
+            )
+
+    return references
```

Comment:
this can be shortened a lot (see comment below): 

just check whether reference_dist is None, then construct a min max dist with correct shapes.

if it is already a Distribution, sample once and assert that it matches theta shape (this could happen in the main tarp function as well). You need just a single assert for this.

Reply:
the use of a distribution object, didn't occur to me. I could go into that direction, but I think this is a detail that is easy to fix later. I removed some shape checks with 90bb7705